### ISSUE TITLE
Tracking the safety classification of the current location of a settler

### ISSUE THREAD

**Issue by Urwumpe** [mars-sim/mars-sim#790]
For further getting the circumstances of Death (or accidents) defined, we need some more knowledge about where and why the accident happened.

For realizing this, a log of the past reason(s) why somebody left and entered the "safe green zone" of the base would be good, like signing on/off in a restricted work place. I suggest extending the Unit window with a table with the following lines:

1. When did the settler leave
2. Where did the settler leave (airlock of building, rover, base)
3. Why did the settler leave (Task, Mission, Order)
4. When did the settler return from this?
5. Where did the settler enter a similar safe or safer zone again?

And for that we would need an enum of a classification of the safety:
- Green (Intact Base: Shirt-sleeves environment)
- Yellow (Rover, limited life support endurance of base: Still breathable, but pressure suits should be donned and ready to be used)
- Red (outside, critically failed life support of a base or rover, pressure suit mandatory, helmets closed)
- Purple (radiation hazard, like near a nuclear reactor or outside during a CME or GRB: Time to stay here is limited even with protection)

I think for the sake of simplificity, this scope should be automatically calculated by Settlement, Rover or EVA suit, maybe triggered by certain events to avoid unneccessary calculations. The list of recent transitions should be like a stack then, one new line being in it for each settler being generated, if the settler moves from a safe place to a less safe place or removed if he returns to a place with higher safety than the previous ones (more than one possible). When a line is removed there, the closed line is moved to the log. When the settler dies, the stack is emptied and all unclosed departures moved to the log.

The number of such transitions should be limited to a low number, I think 6 should already be good, this allows having the past information about such departures before the accident without storing too much data. Better would be being able to configure this limit.

**Comment by Urwumpe** [comment]
Yes, but not just containers, but also states of these containers, so if there is a critical situation, a base or rover should automatically be marked as dangerous and the people automatically transitioned to the same place/rover but in a more dangerous state. Like for example, during a radiation event.

That way, we could for example also see, if a safe place actually wasn't safe at that time because of a malfunction. Ideally, we should be able to tell that such a change is caused by a certain malfunction or lack of consumables. 

I think, just having a log of the last places and a single "action" or service changing the location could still not be accurate enough to tell, why somebody died and how we need to improve the AI. And the minimum goal should at least be that the number of settlers can be kept stable by new arrivals replacing losses.

**Comment by mokun** [comment]
Yes we could save the previous state as well. This way, we can compare the previous and current state to get a better picture of what went wrong.

**Comment by mokun** [comment]
> enum of a classification of the safety:

It's a good idea to give an enum color to denote the type of zone that an item/person is at, or even for the type of work a person/bot engages in.

I suppose if a mechanic has to climb up to the top of a building to do an outside roofing inspect, we may mark the job hazard safety level as ORANGE.

For an engineer to go to MD4 to perform repair, the job hazard should be considered RED.
